seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/CoordinatorService.java

```diff
@@ -0,0 +1,416 @@
+     * <p>
+     * This IMap is used to recovery ownedSlotProfilesIMap in JobMaster when a new master node active
+     */
+    private IMap<PipelineLocation, Map<TaskGroupLocation, SlotProfile>> ownedSlotProfilesIMap;
+
+    /**
+     * If this node is a master node
+     */
+    private volatile boolean isMaster = false;
+
+    private final ExecutorService executorService;
+
+    @SuppressWarnings("checkstyle:MagicNumber")
+    public CoordinatorService(@NonNull NodeEngineImpl nodeEngine, @NonNull ExecutorService executorService) {
+        this.nodeEngine = nodeEngine;
+        this.logger = nodeEngine.getLogger(getClass());
+        this.executorService = executorService;
+
+        ScheduledExecutorService masterActiveListener = Executors.newSingleThreadScheduledExecutor();
+        masterActiveListener.scheduleAtFixedRate(() -> checkNewActiveMaster(), 0, 100, TimeUnit.MILLISECONDS);
+    }
+
+    public JobMaster getJobMaster(Long jobId) {
+        return runningJobMasterMap.get(jobId);
+    }
+
+    // On the new master node
+    // 1. If runningJobStateIMap.get(jobId) == null and runningJobInfoIMap.get(jobId) != null. We will do
+    //    runningJobInfoIMap.remove(jobId)
+    //
+    // 2. If runningJobStateIMap.get(jobId) != null and the value equals JobStatus End State. We need new a
+    //    JobMaster and generate PhysicalPlan again and then try to remove all of PipelineLocation and
+    //    TaskGroupLocation key in the runningJobStateIMap.
+    //
+    // 3. If runningJobStateIMap.get(jobId) != null and the value equals JobStatus.SCHEDULED. We need cancel the job
+    //    and then call submitJob(long jobId, Data jobImmutableInformation) to resubmit it.
+    //
+    // 4. If runningJobStateIMap.get(jobId) != null and the value is CANCELING or RUNNING. We need recover the JobMaster
+    //    from runningJobStateIMap and then waiting for it complete.
+    private void initCoordinatorService() {
+        runningJobInfoIMap = nodeEngine.getHazelcastInstance().getMap("runningJobInfo");
+        runningJobStateIMap = nodeEngine.getHazelcastInstance().getMap("runningJobState");
+        runningJobStateTimestampsIMap = nodeEngine.getHazelcastInstance().getMap("stateTimestamps");
+        ownedSlotProfilesIMap = nodeEngine.getHazelcastInstance().getMap("ownedSlotProfilesIMap");
+
+        List<CompletableFuture<Void>> collect = runningJobInfoIMap.entrySet().stream().map(entry -> {
+            return CompletableFuture.runAsync(() -> restoreJobFromMasterActiveSwitch(entry.getKey(), entry.getValue()),
+                executorService);
+        }).collect(Collectors.toList());
+
+        try {
+            CompletableFuture<Void> voidCompletableFuture = CompletableFuture.allOf(
+                collect.toArray(new CompletableFuture[0]));
+            voidCompletableFuture.get();
+        } catch (Exception e) {
+            throw new SeaTunnelEngineException(e);
```

Comment:
> If some job already restore, but another not successed when restore, do we need to cancel all job or do what?

The initCoordinatorService process allows jobs to fail, because there is no guarantee that those jobs will not fail before the JobMaster fails. We need to ensure that there are no unpredictable exceptions in the initCoordinatorService process.